lib/libxdp/xsk.c

```diff
@@ -398,6 +399,32 @@ static enum xdp_attach_mode xsk_convert_xdp_flags(__u32 xdp_flags)
 	return XDP_MODE_NATIVE;
 }
 
+#define MAX_DEV_QUEUE_PATH_LEN 64
+
+static int xsk_get_queues_from_sysfs(const char* ifname) {
+	DIR *dir;
+	struct dirent *entry;
+	char buf[MAX_DEV_QUEUE_PATH_LEN];
+	int rx = 0, tx = 0;
+
+	snprintf(buf, MAX_DEV_QUEUE_PATH_LEN, "/sys/class/net/%s/queues/", ifname);
+	dir = opendir(buf);
+	if(dir == NULL)
+		return 1;
+
+	while((entry = readdir(dir))) {
+		if (0 == strncmp(entry->d_name, "rx", 2))
+			rx++;
+
+		if (0 == strncmp(entry->d_name, "tx", 2))
+			tx++;
+	}
+
+	closedir(dir);
+
+	return max(max(rx, tx), 1);
```

Comment:
I think that this logic belongs to the calling function. This function should only do what it states: get_queues_from_sysfs (good name BTW) and if it cannot, return an error. Let the logic above set channels to 1 if all methods fail.